### PR TITLE
Add erlang:system_flag(system_logger, _)

### PR DESCRIPTION
**Pull request by garazdawi**
This flag allows logger and other components to set the process which log messages from ERTS are to be sent.

### CODE REVIEW

**Comment by garazdawi** [comment]
We've implemented it because we needed a way in `logger` to not send the emulator messages to the same process all the time. We do not expect it to be very useful to any other logging framework, but have made it a public API because it makes it easier to work with and others may be interested in experimenting with it.

We will be using it to avoid making the `logger` process have to have overload protection from when the emulator generates huge amounts of error messages.

**Comment by josevalim** [comment]
@garazdawi I see, thanks! In this case, maybe we should add some disclaimers (alert boxes) to the docs? Or do you believe it is not necessary?

Also have you considered moving the handling of the emulator messages to a separate `system_logger` process whose sole responsibility is to handle those messages? This means you could implement overload protection but at least in a way it wouldn't affect the main logger process.

**Comment by garazdawi** [comment]
> I see, thanks! In this case, maybe we should add some disclaimers (alert boxes) to the docs? Or do you believe it is not necessary?

I'll see to adding a note.

> Also have you considered moving the handling of the emulator messages to a separate system_logger process whose sole responsibility is to handle those messages? This means you could implement overload protection but at least in a way it wouldn't affect the main logger process.

Yes we've considered it, but I believe this is a better solution as it can handle crashes in the system_logger process without bringing the whole system down.

**Comment by vans163** [comment]
TY finally!

@josevalim The major use case is imagine you have 10,000,000 gen_servers, by default gen_servers when they crash log to the system error logger with a nice red message + stacktrace. No matter what, you cannot turn this behaviour off.

Now imagine what happens if 0.1% of your genservers crash every second. You end up with an unreadable REPL which is getting spammed with red system logger crash errors.

With this patch for that particular gen_server youl be able to unset the system_logger, to prevent this system log spam.

**Comment by garazdawi** [comment]
@vans163 Not sure how this functionality would help with that, as it is proc_lib/gen_server that generate those log messages, not the emulator. However, can't you just use a logger filter for that? Something like this:

```
Filter =
        fun(#{ msg := {report, #{ label := Label }}}, State) ->
                case lists:member(Label, State) of
                    true ->
                        stop;
                    false ->
                        ignore
                end;
           (_Log,_State) ->
                ignore
        end,
    logger:add_primary_filter(gen_server_crash, {Filter, [{gen_server,terminate},{proc_lib,crash}]).
```

**Comment by vans163** [comment]
@garazdawi Bummer.  Yes I am aware but those filters get expensive.  I was hoping there could be a way to turn off the gen_server bottleneck logging, where it spams every crash to logger no matter what (even if you redirect all process IO to another group_leader, the error message instead of going to the group leader goes to the error logger).

**Comment by sirihansen** [comment]
@vans163 If you don't want any gen_server or proc_lib reports, you can set the module level for these modules to 'none'. This will, however, turn off these reports from all gen_server/proc_lib processes in your system - don't know if that's what you're after? (the same would of course be the case with the filter that @garazdawi suggests).

Setting the system_logger to undefined would turn of logging from erts (for example error exceptions in "naked processes") and from remote nodes when the remote process' group_leader is on "your node". Note that the system_logger is a system flag, meaning that it is also set for the system, not per process or module.

**Comment by vans163** [comment]
@sirihansen exactly the problem. I do not want to turn it off for all genservers but for just the particular processes, example a particular module.  So far there seems to be no clean way to do this.

**Comment by garazdawi** [comment]
What about putting a value in the process dictionary and then filtering on that? Filters are run in the process that is doing the logging.

**Comment by sirihansen** [comment]
@vans163 The clean way to do this, still using gen_servers, is as @garazdawi suggets - where "putting a value in the process dictionary" is best done by calling `logger:set_process_metadata/1`, and then add a filter that checks the metadata for whatever you inserted there.

> BTW proc_lib avoids the system_logger. And is the way to avoid this behaviour, to spawn processes using proc_lib instead of using gen_servers or regular processes.

Both proc_lib and gen_server produce error messages when the process crashes (i.e. terminates with a reason that is not `normal`, `shutdown`, or `{shutdown,Term}`). gen_server logs an error report, and proc_lib logs a crash report. Prior to OTP-21, the proc_lib crash report would only be logged if the sasl application was running, but since OTP-21 it is logged by default. Setting kernel confiuration parameter `logger_sasl_compatible` to `true` will make OTP-21 behave as earlier releases when it comes to this. The recommended way is, however, to use filters if the log events are not desired. Alternatively, make sure that your process always terminates with reason `normal`, `shutdown`, or `{shutdown,Term}`.

If you use plain processes (not gen_server or proc_lib), erts will produce an error report if the process terminates with an error exception, but not if it terminates with an exit.

**Comment by vans163** [comment]
@galarsson @sirihansen, Right I see its a new feature added in OTP21, I will play with it, thanks.

The crash report (proc_lib) if the process is part of a trapping exits process, is sent to the parent, instead of the error logger, thus no console spam.

The error report, if the process is part of a trapping exits process or not, is ALWAYS sent to the error / logger and spams console.

File: erts/doc/src/erlang.xml [hunk 1]
```diff
@@ -9057,18 +9090,24 @@ ok
...
             <c>system_architecture</c></tag>
           <item>
             <p>Returns a string containing the processor and OS
               architecture the emulator is built for.</p>
           </item>
+          <tag><marker id="system_info_system_logger"/>
+            <c>system_logger</c></tag>
+          <item>
+            <p>Returns the current <c>system_logger</c> as set by
+            <seealso marker="#process_flag/2"><c>erlang:system_flag(system_logger, _)</c></seealso>.</p>
```
Comment (sirihansen) [inline erts/doc/src/erlang.xml]: Is this marker correct?

File: erts/preloaded/src/erlang.erl [hunk 2]
```diff
@@ -2526,6 +2526,8 @@ subtract(_,_) ->
                                 OldSchedulersOnline when
       SchedulersOnline :: pos_integer(),
       OldSchedulersOnline :: pos_integer();
+                        (system_logger, Logger) -> true when
+      Logger :: logger | pid();
```
Comment (sirihansen) [inline erts/preloaded/src/erlang.erl]: or `undefined`

File: erts/preloaded/src/erlang.erl [hunk 3]
```diff
@@ -2731,8 +2733,9 @@ tuple_to_list(_Tuple) ->
          (schedulers | schedulers_online) -> pos_integer();
          (smp_support) -> boolean();
          (start_time) -> integer();
-         (system_version) -> string();
          (system_architecture) -> string();
+         (system_logger) -> logger | pid();
```
Comment (sirihansen) [inline erts/preloaded/src/erlang.erl]: or `undefined`

**Pull request merged**